### PR TITLE
feat(filesystem): implement MCP roots protocol for dynamic directory management

### LINKED ISSUE #401: Support "roots" in filesystem MCP server
**Is your feature request related to a problem? Please describe.**
Currently, the filesystem server only allows a fixed set of allowed directories specified via an arg.

**Describe the solution you'd like**
The filesystem server could leverage "roots" to allow the client to update the list of directories that the server can operate on.

**Describe alternatives you've considered**
None

**Additional context**
See this discussion: https://github.com/eclipse-theia/theia/discussions/14655#discussioncomment-11632301

### PR DESCRIPTION
**Pull request by nandsha**
## Description
Implements the MCP roots protocol in the filesystem server to enable dynamic directory access management. This allows clients to update the list of accessible directories at runtime without restarting the server.

## Server Details
- Server: filesystem
- Changes to: core directory access control, initialization flow, documentation

## Motivation and Context
Fixes #401 - The filesystem server previously had a fixed set of allowed directories specified via command-line arguments. This implementation adds support for the MCP roots protocol, enabling flexible directory access configuration and runtime updates to accessible filesystem paths.

## How Has This Been Tested?
- Tested with roots-capable client - MCP Inspector (successful dynamic directory updates)
- Tested for backwards compatibility with args
- Tested edge cases: empty roots, invalid directories, no initial directories
- All existing tests pass
- Added test coverage for roots processing utility

## Breaking Changes
No breaking changes. The implementation maintains full backwards compatibility:
- Existing command-line configurations continue to work unchanged
- Clients without roots support see no change in behaviour
- Only change: roots-capable clients can now override server directories

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist
- [x] I have read the [MCP Protocol Documentation](https://modelcontextprotocol.io)
- [x] My changes follows MCP security best practices
- [x] I have updated the server's README accordingly
- [x] I have tested this with an LLM client
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have documented all environment variables and configuration options

### Implementation Details:
- Server checks client capabilities during initialization
- Roots completely replace command-line directories (not additive)
- Server fetches roots on two occasions:
  1. After initialization (via `oninitialized` callback)
  2. When client sends `roots/list_changed` notification
- Comprehensive error handling for missing directories scenario

### Key Changes:
1. **index.ts**: 
   - Added roots protocol support with `oninitialized` callback
   - Added notification handler for `roots/list_changed`
   - Enhanced error messages for better user guidance

2. **README.md**: 
   - Added "Directory Access Control" section explaining both methods
   - Documented the interaction between command-line args and MCP roots

3. **roots-utils.ts:**
  - Added roots processing logic as testable utility module

4. **__tests__/roots-utils.test.ts:
  - Added test coverage for all URI formats (file://, plain paths)
  - Tests for error scenarios (non-existent dirs, files vs directories)
  - Edge case testing (empty arrays, invalid paths, complex path normalization)
  - 100% statement and function coverage

### CODE REVIEW

**Comment by olaservo** [comment]
Hi @nandsha , thanks a lot for adding this!  I finally merged in some changes today to support adding and running tests for the filesystem server: https://github.com/modelcontextprotocol/servers/pull/1405

Could you get the latest from main and add some test coverage for these changes?

**Comment by nandsha** [comment]
Sure @olaservo, thanks for adding the tests and reaching out

**Comment by nandsha** [comment]
Hi @olaservo, refactored roots processing logic into `roots-utils.ts` to make it testable, added full test coverage, and it's ready for your review. Please take a look. Thanks!

File: src/filesystem/index.ts [hunk 1]
```diff
@@ -839,12 +845,59 @@ server.setRequestHandler(CallToolRequestSchema, async (request) => {
...
+server.oninitialized = async () => {
+  const clientCapabilities = server.getClientCapabilities();
+
+  if (clientCapabilities?.roots) {
+    try {
+      const response = await server.listRoots();
+      if (response && 'roots' in response) {
+        await updateAllowedDirectoriesFromRoots(response.roots);
+      } else {
+        console.log("Client returned no roots set, keeping current settings");
```
Comment (olaservo) [inline src/filesystem/index.ts]:
```suggestion
        console.error("Client returned no roots set, keeping current settings");
```

(Since server uses stdio transport I think we need to keep these all stderr)

File: src/filesystem/index.ts [hunk 2]
```diff
@@ -839,12 +845,59 @@ server.setRequestHandler(CallToolRequestSchema, async (request) => {
...
+      }
+    } catch (error) {
+      console.error("Failed to request initial roots from client:", error instanceof Error ? error.message : String(error));
+    }
+  } else {
+    if (allowedDirectories.length > 0) {
+      console.log("Client does not support MCP Roots, using allowed directories set from server args:", allowedDirectories);
```
Comment (olaservo) [inline src/filesystem/index.ts]:
```suggestion
      console.error("Client does not support MCP Roots, using allowed directories set from server args:", allowedDirectories);
```

File: src/filesystem/index.ts [hunk 3]
```diff
@@ -839,12 +845,59 @@ server.setRequestHandler(CallToolRequestSchema, async (request) => {
...
+};
+
 // Start server
 async function runServer() {
   const transport = new StdioServerTransport();
   await server.connect(transport);
   console.error("Secure MCP Filesystem Server running on stdio");
-  console.error("Allowed directories:", allowedDirectories);
+  if (allowedDirectories.length === 0) {
+    console.log("Started without allowed directories - waiting for client to provide roots via MCP protocol");
```
Comment (olaservo) [inline src/filesystem/index.ts]:
```suggestion
    console.error("Started without allowed directories - waiting for client to provide roots via MCP protocol");
```

File: src/filesystem/README.md [hunk 4]
```diff
@@ -9,8 +9,58 @@ Node.js server implementing Model Context Protocol (MCP) for filesystem operatio
 - Move files/directories
 - Search files
 - Get file metadata
+- Dynamic directory access control via MCP roots protocol
+
+## Directory Access Control
+
+The server uses a flexible directory access control system. Directories can be specified via command-line arguments or dynamically via the MCP roots protocol.
```
Comment (olaservo) [inline src/filesystem/README.md]:
```suggestion
The server uses a flexible directory access control system. Directories can be specified via command-line arguments or dynamically via [Roots](https://modelcontextprotocol.io/docs/concepts/roots).
```

File: src/filesystem/README.md [hunk 5]
```diff
@@ -9,8 +9,58 @@ Node.js server implementing Model Context Protocol (MCP) for filesystem operatio
 - Move files/directories
 - Search files
 - Get file metadata
+- Dynamic directory access control via MCP roots protocol
```
Comment (olaservo) [inline src/filesystem/README.md]:
```suggestion
- Dynamic directory access control via [Roots](https://modelcontextprotocol.io/docs/concepts/roots)
```

This also might be redundant to the line below?

File: src/filesystem/README.md [hunk 6]
```diff
@@ -9,8 +9,58 @@ Node.js server implementing Model Context Protocol (MCP) for filesystem operatio
...
+
+The server uses a flexible directory access control system. Directories can be specified via command-line arguments or dynamically via the MCP roots protocol.
+
+### Method 1: Command-line Arguments
+Specify Allowed directories when starting the server:
+```bash
+mcp-server-filesystem /path/to/dir1 /path/to/dir2
+```
+
+### Method 2: MCP Roots Protocol (Recommended)
```
Comment (olaservo) [inline src/filesystem/README.md]:
```suggestion
### Method 2: MCP Roots (Recommended)
```

File: src/filesystem/README.md [hunk 7]
```diff
@@ -9,8 +9,58 @@ Node.js server implementing Model Context Protocol (MCP) for filesystem operatio
...
+MCP clients that support the roots protocol can dynamically update the Allowed directories.
```
Comment (olaservo) [inline src/filesystem/README.md]:
```suggestion
MCP clients that support [Roots](https://modelcontextprotocol.io/docs/concepts/roots) can dynamically update the Allowed directories. 
```

File: src/filesystem/README.md [hunk 8]
```diff
@@ -9,8 +9,58 @@ Node.js server implementing Model Context Protocol (MCP) for filesystem operatio
...
+```
+
+### Method 2: MCP Roots Protocol (Recommended)
+MCP clients that support the roots protocol can dynamically update the Allowed directories. 
+
+Roots notified by Client to Server, completely replace any server-side Allowed directories when provided.
+
+**Important**: If server starts without command-line arguments AND client doesn't support roots protocol (or provides empty roots), the server will throw an error during initialization.
+
+This is the recommended method, as MCP roots protocol for dynamic directory management. This enables runtime directory updates via `roots/list_changed` notifications without server restart, providing a more flexible and modern integration experience.
```
Comment (olaservo) [inline src/filesystem/README.md]:
```suggestion
This is the recommended method, as this enables runtime directory updates via `roots/list_changed` notifications without server restart, providing a more flexible and modern integration experience.
```

**Review by olaservo — CHANGES_REQUESTED** [review]
Left a few comments on docs and logging, otherwise I think this LGTM.

**Comment by nandsha** [comment]
Thanks again @olaservo! Applied the updates from your suggestions. Appreciate you taking the time!

**Comment by olaservo** [comment]
@nandsha Sorry one quick ask - can you pull in the latest from main?  We merged in a couple updates to CI that I'd like to run on this PR.  Thanks!

**Comment by nandsha** [comment]
All good @olaservo, rebased and ready!

**Comment by olaservo** [comment]
Thanks @nandsha , LGTM!  Heads up I'm still checking in on another PR from the core team that might need to go ahead of this, but I'll plan to merge this one on Monday.

**Comment by nandsha** [comment]
Hi @olaservo, I have synced the latest changes and added symlink resolution and home directory expansion support for roots URI processing. Thanks!

**Comment by olaservo** [comment]
Hi @nandsha , I checked this out locally on my Windows machine since I realized that platform differences can have a big impact with this server.  I ran into some issues running the race condition tests on Windows, that were due to the environment being less permissive with creating symlinks by default. There is also a Windows-specific issue with the way the path is being validated.  However, there is also an incoming PR I'm helping prepare to add Windows fixes, so I am going to go ahead and merge this and then address those as a fast follow.

One other thing I'm thinking about is how to validate this more fully end-to-end with a client that is actually integrated with a model.  I think that is outside of the scope of the groundwork you did here, since this change has been validated with Inspector.  Basically this PR is still ahead of the curve on where most host apps are at wrt Roots (to the best of my knowledge).  But I think it will actually be useful in helping us solve that chicken or egg problem.

If you have any interest in getting involved with a new Community Working Group around validation and reference implementations, please take a look at the CWG repo for more info on how to get involved there too: https://github.com/modelcontextprotocol-community/working-groups?tab=readme-ov-file#participation-and-contributions

**Review by olaservo — APPROVED** [review]

**Comment by therealsickonechase-bit** [comment]
No directory root
